pkg/storage/stores/tsdb/index/index.go

```diff
@@ -2088,57 +2203,294 @@ func buildChunkSamples(d encoding.Decbuf, numChunks int, info *chunkSamples) err
+
+			if overlap(from, through, chunkMeta.MinTime, chunkMeta.MaxTime) {
+				// add to stats
+				res.AddChunk(chunkMeta, from, through)
+			} else if chunkMeta.MinTime >= through {
+				break
+			}
+		}
+	}
+
+	return res, d.Err()
+
+}
+
+func (dec *Decoder) accumulateChunkStats(d *encoding.Decbuf, nChunks int, from, through int64) (res ChunkStats, err error) {
+	var prevMaxT int64
+	for i := 0; i < nChunks; i++ {
+		chunkMeta := &ChunkMeta{}
+		if err := readChunkMeta(d, prevMaxT, chunkMeta); err != nil {
+			return res, errors.Wrap(d.Err(), "read meta for chunk")
+		}
+		prevMaxT = chunkMeta.MaxTime
+
+		if overlap(from, through, chunkMeta.MinTime, chunkMeta.MaxTime) {
+			// add to stats
+			res.AddChunk(chunkMeta, from, through)
+		} else if chunkMeta.MinTime >= through {
+			break
+		}
+	}
+	return res, d.Err()
+}
+
+func (dec *Decoder) readChunkStatsPriorV3(d *encoding.Decbuf, seriesRef storage.SeriesRef, from, through int64) (res ChunkStats, err error) {
+	// prior to v3, chunks needed iteration for stats aggregation
+	chks := ChunkMetasPool.Get()
+	defer ChunkMetasPool.Put(chks)
+	err = dec.readChunks(FormatV2, d, seriesRef, from, through, &chks)
+	if err != nil {
+		return ChunkStats{}, err
+	}
+
+	for _, chk := range chks {
+		if overlap(from, through, chk.MinTime, chk.MaxTime) {
+			res.AddChunk(&chk, from, through)
+		} else if chk.MinTime >= through {
+			break
+		}
+
+	}
+
+	return res, nil
+
+}
+
+// Series decodes a series entry from the given byte slice into lset and chks.
+func (dec *Decoder) Series(version int, b []byte, seriesRef storage.SeriesRef, from int64, through int64, lbls *labels.Labels, chks *[]ChunkMeta) (uint64, error) {
+
+	d, fprint, err := dec.prepSeries(b, lbls, chks)
+	if err != nil {
+		return 0, err
+	}
+
+	// read chunks based on fmt
+	if err := dec.readChunks(version, d, seriesRef, from, through, chks); err != nil {
+		return 0, err
+	}
+	return fprint, nil
+
+}
+
+func (dec *Decoder) readChunks(version int, d *encoding.Decbuf, seriesRef storage.SeriesRef, from int64, through int64, chks *[]ChunkMeta) error {
+	// read chunks based on fmt
+	if version > FormatV2 {
+		return dec.readChunksV3(d, from, through, chks)
+	}
+	return dec.readChunksPriorV3(d, seriesRef, from, through, chks)
+}
+
+func (dec *Decoder) readChunksV3(d *encoding.Decbuf, from int64, through int64, chks *[]ChunkMeta) error {
+	nChunks := d.Uvarint()
+	chunksRemaining := nChunks
+
+	markersLn := int(d.Be32()) // markersLn
+
+	// variables must be declared before goto, allowing us to skip
+	// using chunk pages when the chunk count is small
+	var (
+		nMarkers int
+		marker   chunkPageMarker
+		markers  chunkPageMarkers
+		prevMaxT int64
+	)
+
+	startMarkers := d.Len()
+	if nChunks < dec.maxChunksToBypassMarkerLookup {
+		d.Skip(markersLn)
+		goto iterate
+	}
+
+	nMarkers = d.Uvarint()
+
+	// TODO(owen-d): use pool
+	markers = make(chunkPageMarkers, nMarkers)
+	for i := range markers {
+		markers[i].decode(d)
```

Comment:
I was wondering how `d` was reset back to the position where the next page marker lies after it being used to read the chunks below, and it seems we just continue reading the chunks from the first overlapping page until we hit the loop termination or `chunkMeta.MinTime >= through` condition. Also, as soon as the loop ends, we immediately return from the function since the runtime doesn't jump back to where the goto was called. See [this example](https://go.dev/play/p/OYR01Di6Ogq).

The tests are passing because we go through all the chunks starting with chunks from first overlapping page.

Reply:
Having some trouble understanding your comment, but I'll try to explain:

We iterate through decoding pages until we find a page that overlaps with the time range we want. Upon finding a page that has data we need, we skip the the beginning of that page. At this point, we won't care about the page markers anymore; we'll just iterate through chunks directly. We don't ever need to return to the first loop.